src/storage/QueueStorage.ts

```diff
@@ -0,0 +1,112 @@
+import { Hash, ForkId, BlockHeight } from "@/types/types";
+
+export class QueueStorage {
+    private queuedBlocks: Map<Hash, BlockConfirmationStruct> = new Map();
+
+    // Secondary index for efficient queries by coordinates
+    private blocksByCoordinates: Map<string, Set<Hash>> = new Map();
+
+    /** Queue a block for future processing */
+    queueBlock(signedBlock: SignedBlockStruct): Hash {
+        const block = Block.decode(signedBlock.encodedBlock);
+
+        // Convert to BlockConfirmationStruct with empty signatures
+        const blockConfirmation: BlockConfirmationStruct = {
+            signedBlock: signedBlock,
+            signatures: []
+        };
+
+        this.queuedBlocks.set(block.hash, blockConfirmation);
+
+        // Update coordinate index
+        this.addHashToCoordinateIndex(block.hash, block.forkId, block.height);
+
+        return block.hash;
+    }
+
+    /** Queue a block confirmation for future processing */
+    queueConfirmation(blockConfirmation: BlockConfirmationStruct): Hash {
+        const block = Block.decode(blockConfirmation.signedBlock.encodedBlock);
+
+        // Check if block already exists in queue
+        const existingBlockConfirmation = this.queuedBlocks.get(block.hash);
+
+        if (existingBlockConfirmation) {
+            // Merge signatures from the new confirmation
+            const signaturesSet = new Set(existingBlockConfirmation.signatures);
+            for (const newSignature of blockConfirmation.signatures) {
+                signaturesSet.add(newSignature);
+            }
+
+            existingBlockConfirmation.signatures = Array.from(signaturesSet);
+
+            return block.hash;
+        }
+        // Store the new block confirmation
+        this.queuedBlocks.set(block.hash, blockConfirmation);
+        this.addHashToCoordinateIndex(block.hash, block.forkId, block.height);
+
+        return block.hash;
+    }
+
+    /** Try to dequeue confirmations for a specific fork/height */
+    tryDequeueConfirmations(
+        forkId: ForkId,
+        height: BlockHeight
+    ): BlockConfirmationStruct[] {
+        const coordinateKey = this.coordinatesToKey(forkId, height);
+        const hashSet = this.blocksByCoordinates.get(coordinateKey);
+
+        if (!hashSet) {
+            return [];
+        }
+
+        const blockConfirmations: BlockConfirmationStruct[] = [];
+
+        // Collect all blocks for this coordinate
+        for (const hash of hashSet) {
+            const blockConfirmation = this.queuedBlocks.get(hash);
+            if (blockConfirmation) {
+                blockConfirmations.push(blockConfirmation);
+
+                // delete the block from the queue
+                this.queuedBlocks.delete(hash);
+            }
+        }
+
+        // Remove from queue
+        this.blocksByCoordinates.delete(coordinateKey);
+
+        return blockConfirmations;
+    }
+
+    isBlockQueued(hash: Hash): boolean {
+        return this.queuedBlocks.has(hash);
```

Comment:
this is good, but in general where it's used better to still try and persist the `BlockConfirmation` so it potentially expands the `signature set`

Reply:
to consider `isBlockQueues(blockConfirmation, option{hash?}){
if(not in hash map) return false;
else try expand signatures;
return true;
}

Reply:
actually not even used => deleted